File: tutorial/search/binarySearch/BinarySearch.java
import java.io.*;
import java.util.*;;

class BinarySearch {
    public static void main(String[] args) {
        int[] list = { 1, 2, 3, 4, 5, 6, 7, 8 };
        int key = 5;
        System.out.println(BSearch(list, key));

    }

    static int BSearch(int list[], int key) {

        Arrays.sort(list);
        int left = 0;
        int right = list.length - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (key == list[mid])
                return mid;
            if (key < list[mid])
                left = mid + 1;
            if (key > list[mid])
                right = mid - 1;
        }
        return -1;

    }
}

File: tutorial/search/binarySearch/BinarySearch.go
package main

import "fmt"

func BSearch_Recursion(list []int, key int) int {
	left := 0
	right := len(list) - 1
	for left <= right {
		mid := (left + right) / 2
		if key == list[mid] {
			return mid
		} else if list[mid] < key {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}
	return -1
}

func main() {
	list := []int{1, 2, 4, 5, 6}
	fmt.Println(BSearch_Recursion(list, 5))
}
